src/pymor/reductors/era.py

```diff
@@ -0,0 +1,246 @@
+        return mp
+
+    @cached
+    def _sv_U_V(self, num_left, num_right):
+        n, p, m = self.data.shape
+        s = n if self.force_stability else (n + 1) // 2
+        if num_left is None and m * s < p:
+            self.logger.info('Data has low rank! Accelerating computation with output tangential projections ...')
+            num_left = m * s
+        if num_right is None and p * s < m:
+            self.logger.info('Data has low rank! Accelerating computation with input tangential projections ...')
+            num_right = p * s
+        h = self._project_markov_parameters(num_left, num_right) if num_left or num_right else self.data
+        self.logger.info(f'Computing SVD of the {"projected " if num_left or num_right else ""}Hankel matrix ...')
+        if self.force_stability:
+            h = np.concatenate([h, np.zeros_like(h)[1:]], axis=0)
+        U, sv, V = spla.svd(to_matrix(NumpyHankelOperator(h)), full_matrices=False)
+        return sv, U.T, V
+
+    def output_projector(self, num_left):
+        """Construct the left/output projector :math:`W_1`."""
+        assert isinstance(num_left, int) and num_left <= self.data.shape[1]
+        self.logger.info(f'Constructing output projector ({num_left} tangential directions) ...')
+        return self._s1_W1()[1][:, :num_left]
+
+    def input_projector(self, num_right):
+        """Construct the right/input projector :math:`W_2`."""
+        assert isinstance(num_right, int) and num_right <= self.data.shape[2]
+        self.logger.info(f'Constructing input projector ({num_right} tangential directions) ...')
+        return self._s2_W2()[1][:, :num_right]
+
+    def error_bounds(self, num_left=None, num_right=None):
+        r"""Compute the error bounds for all possible reduction orders.
+
+        Without tangential projection of the Markov parameters, the error bounds are given by
+
+        .. math::
+            \sum_{i = 1}^{2 s - 1}
+            \lVert C_r A_r^{i - 1} B_r - h_i \rVert_F^2
+            \leq
+            \sigma_{r + 1}(\mathcal{H})
+            \sqrt{r + p + m},
+
+        where :math:`(A_r,B_r,C_r)` is the reduced realization of order :math:`r`,
+        :math:`h_i\in\mathbb{R}^{p\times m}` is the :math:`i`-th Markov parameter
+        and :math:`\sigma_{r+1}(\mathcal{H})` is the first neglected singular value of the
+        Hankel matrix of Markov parameters.
+
+        With tangential projection, the bound is given by
+
+        .. math::
+            \sum_{i = 1}^{2 s - 1}
+            \lVert C_r A_r^{i - 1} B_r - h_i \rVert_F^2
+            \leq
+            4 \left(
+                \sum_{i = n_L + 1}^p \sigma_i^2(\Theta_L)
+                + \sum_{i = n_R + 1}^m \sigma_i^2(\Theta_R)
+            \right)
+            + 2 \sigma_{r + 1}(\mathcal{H}) \sqrt{r + n_L + n_R},
+
+        where :math:`\Theta_L,\,\Theta_R` is the matrix of horizontally or vertically stacked Markov
+        parameters, respectively. See :cite:`KG16` (Thm. 3.4) for details.
```

Comment:
It would be good to put square roots in the math equations, or somehow explain what error bounds are computed.

Reply:
Like so?

Reply:
It seems like you used a strange UTF character that only looks like ^, but is not actually ^. Otherwise, looks ok.

Reply:
Darnit. Apple US Intl. Keyboard...